File: starter_kits/TypeScript/MyBot.ts
import { Constants } from "./hlt/Constants";
import { Direction } from "./hlt/Direction";
import { Game } from "./hlt/Game";
import { Logging } from "./hlt/Logging";
import { Random } from "./hlt/Random";

const firstArgument = process.argv[2];
const random = new Random(firstArgument ? Number(firstArgument) : 324231);

const game = new Game();
game.initialize().then(async ([gameMap, me]) => {
    // At this point "game" variable is populated with initial map data.
    // This is a good place to do computationally expensive start-up pre-processing.
    // As soon as you call "ready" function below, the 2 second per turn timer will start.
    await game.ready("MyTypeScriptBot");

    Logging.info(`My Player ID is ${game.myId}.`);

    while (true) {
        await game.updateFrame();

        const commandQueue = [];

        // Movement algorithm parameters
        const fullCargoCoefficient = 0.5;
        const tooLittleHaliteToKeepHarvestingCoefficient = 0.1;
        const spawnNewShipsUntilGameTurnCoefficient = 0.75;

        for (const ship of me.getShips()) {
            if (ship.haliteAmount > fullCargoCoefficient * Constants.MAX_ENERGY) {
                // Go home if full
                const destination = me.shipyard.position;
                const safeMove = gameMap.naiveNavigate(ship, destination);
                commandQueue.push(ship.move(safeMove));
            } else if (
                gameMap.get(ship.position).haliteAmount <
                tooLittleHaliteToKeepHarvestingCoefficient * Constants.MAX_ENERGY
            ) {
                // Go harvest to a new place
                const getRandomArrayElement = (anArray: any[]) => anArray[Math.floor(anArray.length * random.next())];
                const direction = getRandomArrayElement(Direction.getAllCardinals());
                const destination = ship.position.directionalOffset(direction);
                const safeMove = gameMap.naiveNavigate(ship, destination);
                commandQueue.push(ship.move(safeMove));
            } else {
                // Keep still and harvest in place
            }
        }

        if (game.turnNumber < spawnNewShipsUntilGameTurnCoefficient * Constants.MAX_TURNS &&
            // Spawn new ships
            me.haliteAmount >= Constants.SHIP_COST &&
            !gameMap.get(me.shipyard).isOccupied) {
            commandQueue.push(me.shipyard.spawn());
        }

        await game.endTurn(commandQueue);
    }
});

File: starter_kits/Go/src/MyBot.go
package main

import (
	"fmt"
	"math/rand"
	"os"
	"os/signal"
	"strconv"
	"syscall"
	"time"

	"./hlt"
	"./hlt/gameconfig"
	"./hlt/log"
)

func gracefulExit(logger *log.FileLogger) {
	var gracefulStop = make(chan os.Signal)
	signal.Notify(gracefulStop, syscall.SIGTERM)
	signal.Notify(gracefulStop, syscall.SIGINT)
	go func() {
		sig := <-gracefulStop
		fmt.Printf("caught sig: %+v", sig)
		fmt.Println("Wait for 2 second to finish processing")
		time.Sleep(2 * time.Second)
		logger.Close()
		os.Exit(0)
	}()
}

func main() {
	args := os.Args
	var seed = time.Now().UnixNano() % int64(os.Getpid())
	if len(args) > 1 {
		seed, _ = strconv.ParseInt(args[1], 10, 64)
	}
	rand.Seed(seed)

	var game = hlt.NewGame()
	// At this point "game" variable is populated with initial map data.
	// This is a good place to do computationally expensive start-up pre-processing.
	// As soon as you call "ready" function below, the 2 second per turn timer will start.

	var config = gameconfig.GetInstance()
	fileLogger := log.NewFileLogger(game.Me.ID)
	var logger = fileLogger.Logger
	logger.Printf("Successfully created bot! My Player ID is %d. Bot rng seed is %d.", game.Me.ID, seed)
	gracefulExit(fileLogger)
	game.Ready("MyBot")
	for {
		game.UpdateFrame()
		var me = game.Me
		var gameMap = game.Map
		var ships = me.Ships
		var commands = []hlt.Command{}
		for i := range ships {
			var ship = ships[i]
			var maxHalite, _ = config.GetInt(gameconfig.MaxHalite)
			var currentCell = gameMap.AtEntity(ship.E)
			if currentCell.Halite < (maxHalite/10) || ship.IsFull() {
				commands = append(commands, ship.Move(hlt.AllDirections[rand.Intn(4)]))
			} else {
				commands = append(commands, ship.Move(hlt.Still()))
			}
		}
		var shipCost, _ = config.GetInt(gameconfig.ShipCost)
		if game.TurnNumber <= 200 && me.Halite >= shipCost && !gameMap.AtEntity(me.Shipyard.E).IsOccupied() {
			commands = append(commands, hlt.SpawnShip{})
		}
		game.EndTurn(commands)
	}
}
